agora/Agora/Governor.hs

```diff
@@ -39,30 +76,112 @@ PlutusTx.makeIsDataIndexed ''GovernorDatum [('GovernorDatum, 0)]
-governorPolicy :: Governor -> ClosedTerm PMintingPolicy
-governorPolicy _ =
-  plam $ \_redeemer _ctx' -> P.do
-    popaque (pconstant ())
-
--- | Validator for Governors.
-governorValidator :: Governor -> ClosedTerm PValidator
-governorValidator _ =
-  plam $ \_datum _redeemer _ctx' -> P.do
-    popaque (pconstant ())
+-- | Plutarch-level datum for the Governor script.
+newtype PGovernorDatum (s :: S) = PGovernorDatum
+  { getGovernorDatum ::
+    Term
+      s
+      ( PDataRecord
+          '[ "proposalThresholds" ':= PProposalThresholds
+           , "nextProposalId" ':= PProposalId
+           ]
+      )
+  }
+  deriving stock (GHC.Generic)
+  deriving anyclass (Generic)
+  deriving anyclass (PIsDataRepr)
+  deriving
+    (PlutusType, PIsData, PDataFields)
+    via PIsDataReprInstances PGovernorDatum
+
+instance PUnsafeLiftDecl PGovernorDatum where type PLifted PGovernorDatum = GovernorDatum
+deriving via (DerivePConstantViaData GovernorDatum PGovernorDatum) instance (PConstantDecl GovernorDatum)
+
+-- FIXME: derive this via 'PIsDataReprInstances'
+-- Blocked by: PProposalThresholds
+instance PTryFrom PData (PAsData PGovernorDatum) where
+  type PTryFromExcess PData (PAsData PGovernorDatum) = Const ()
+
+  ptryFrom' d k = k (punsafeCoerce d, ())
+
+-- | Plutarch-level version of 'GovernorRedeemer'.
+data PGovernorRedeemer (s :: S)
+  = PCreateProposal (Term s (PDataRecord '[]))
+  | PMintGATs (Term s (PDataRecord '[]))
+  | PMutateGovernor (Term s (PDataRecord '[]))
+  deriving stock (GHC.Generic)
+  deriving anyclass (Generic)
+  deriving anyclass (PIsDataRepr)
+  deriving
+    (PlutusType, PIsData)
+    via PIsDataReprInstances PGovernorRedeemer
+
+instance PUnsafeLiftDecl PGovernorRedeemer where type PLifted PGovernorRedeemer = GovernorRedeemer
+deriving via (DerivePConstantViaData GovernorRedeemer PGovernorRedeemer) instance (PConstantDecl GovernorRedeemer)
+
+deriving via PAsData (PIsDataReprInstances PGovernorRedeemer) instance PTryFrom PData (PAsData PGovernorRedeemer)
+
+--------------------------------------------------------------------------------
+
+-- | Plutrach version of 'getNextProposalId'.
+pgetNextProposalId :: Term s (PProposalId :--> PProposalId)
+pgetNextProposalId = phoistAcyclic $ plam $ \(pto -> pid) -> pcon $ PProposalId $ pid + 1
+
+-- | Get next proposal id.
+getNextProposalId :: ProposalId -> ProposalId
+getNextProposalId (ProposalId pid) = ProposalId $ pid + 1
+
+--------------------------------------------------------------------------------
+
+governorDatumValid :: Term s (PGovernorDatum :--> PBool)
+governorDatumValid = phoistAcyclic $
+  plam $ \datum -> P.do
+    thresholds <-
+      pletFields @'["execute", "draft", "vote"] $
+        pfield @"proposalThresholds" # datum
+
+    execute <- plet $ puntag thresholds.execute
+    draft <- plet $ puntag thresholds.draft
+    vote <- plet $ puntag thresholds.vote
+
+    foldr1
+      (#&&)
+      [ ptraceIfFalse "Execute threshold larger than 0" $ 0 #<= execute
```

Comment:
Think through what `ptraceIfFalse` is for. In the first example, desired behaviour is that the execute threshold _is larger than zero_, so the trace will only appear if the example threshold is _less than or equal to zero_. Therefore the trace should be "Execute threshold is less than or equal to zero", because that's the unintended behaviour we wish to report. 

Does that make sense?